modules/gplazma2-oidc-te/src/main/java/org/dcache/gplazma/tokenx/TokenExchange.java

```diff
@@ -0,0 +1,140 @@
+package org.dcache.gplazma.tokenx;
+
+import static java.util.Objects.requireNonNull;
+import static org.dcache.gplazma.util.Preconditions.checkAuthentication;
+
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.security.Principal;
+import java.util.Set;
+
+import org.apache.http.HttpEntity;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpPost;
+import org.apache.http.client.utils.URIBuilder;
+import org.apache.http.entity.StringEntity;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.util.EntityUtils;
+import org.dcache.auth.BearerTokenCredential;
+import org.dcache.auth.attributes.Restriction;
+import org.dcache.gplazma.AuthenticationException;
+import org.dcache.gplazma.plugins.GPlazmaAuthenticationPlugin;
+import org.dcache.gplazma.util.JsonWebToken;
+import org.json.JSONObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.VisibleForTesting;
+
+
+public class TokenExchange implements GPlazmaAuthenticationPlugin {
+
+    private final static Logger LOG = LoggerFactory.getLogger(TokenExchange.class);
+
+    private final static String TOKEN_EXCHANGE_URL = "gplazma.oidc-te.url";
+    private final static String CLIENT_ID = "gplazma.oidc-te.client-id";
+    private final static String CLIENT_SECRET = "gplazma.oidc-te.client-secret";
+    private final static String GRANT_TYPE = "gplazma.oidc-te.grant-type";
+    private final static String SUBJECT_ISSUER = "gplazma.oidc-te.subject-issuer";
+    private final static String SUBJECT_TOKEN_TYPE = "gplazma.oidc-te.subject-token-type";
+    private final static String AUDIENCE = "gplazma.oidc-te.audience";
+
+    private final CloseableHttpClient client;
+
+    public TokenExchange () {
+        this.client = HttpClients.createDefault();
+    }
+
+    @VisibleForTesting
+    public TokenExchange (CloseableHttpClient client) {
+        this.client = requireNonNull(client);
+    }
+
+    @Override
+    public void authenticate(Set<Object> publicCredentials, Set<Object> privateCredentials,
+        Set<Principal> identifiedPrincipals, Set<Restriction> restrictions)
+        throws AuthenticationException {
+
+        BearerTokenCredential credential = null;
+        for (Object c : privateCredentials) {
+            if (c instanceof BearerTokenCredential) {
+
+                credential = (BearerTokenCredential) c;
+                checkAuthentication(credential == null, "Multiple bearer token credentials");
+
+            }
+        }
+
+        String token = credential.getToken();
+        LOG.debug("Found bearer token: {}", token);
+
+        checkAuthentication(token != null, "No bearer token in the credentials");
+
+        String exchangedToken = null;
```

Comment:
the variable should be defined within try-catch block

Reply:
Thanks, I have moved the variable within the try-catch block.